contracts/Marketplace.sol

```diff
@@ -373,24 +376,43 @@ contract Marketplace is Proofs, StateRetrieval, Endian {
     address withdrawRecipient
   ) public {
     Request storage request = _requests[requestId];
-    require(
-      block.timestamp > requestExpiry(requestId),
-      "Request not yet timed out"
-    );
     require(request.client == msg.sender, "Invalid client address");
     RequestContext storage context = _requestContexts[requestId];
-    require(context.state == RequestState.New, "Invalid state");
+    RequestState state = requestState(requestId);
+    require(
+      state == RequestState.Cancelled ||
+        state == RequestState.Failed ||
+        state == RequestState.Finished,
+      "Invalid state"
+    );
+    require(context.fundsToReturnToClient != 0, "Nothing to withdraw");
+
+    if (state == RequestState.Cancelled) {
+      context.state = RequestState.Cancelled;
+      emit RequestCancelled(requestId);
+
+      // `fundsToReturnToClient` currently tracks funds to be returned for requests that successfully finish.
+      // When requests are cancelled, funds earmarked for payment for the duration
+      // between request expiry and request end (for every slot that was filled), should be returned to the client.
+      // Update `fundsToReturnToClient` to reflect this.
+      context.fundsToReturnToClient +=
+        context.slotsFilled *
+        _payoutAmount(requestId, requestExpiry(requestId));
+    } else if (state == RequestState.Failed) {
+      // For Failed requests the client is refunded whole amount.
+      context.fundsToReturnToClient = request.maxPrice();
```

Comment:
Yeah makes sense. Although I feel like it is a fairly weak mitigation. This mainly could open attacks on the hosts before ending of the request to wiggle out of payment for them... I guess since we are gonna have limited duration for Requests this should be OK as the cost of the attack would be most probably higher than what you would wiggled out from. But maybe the request's duration limit should stay on the lower end "for ever"?

Reply:
Good point. It's not a very practical attack anyway.

I would not be in favor of keeping the duration limit low for ever. It limits the usefulness of the network.

Reply:
> I would not be in favor of keeping the duration limit low for ever. It limits the usefulness of the network.

I disagree about the limiting of the usefulness of the network. IMHO we can nicely plug in account abstraction for "recurring request renewal" to host requests "for ever". If we allow long requests in the system, then we will open doors to this attack that you mentioned before.

Reply:
> If we allow long requests in the system, then we will open doors to this attack that you mentioned before.

30 days is long enough for this attack. A DDOS attack would only need to be long enough that it can force slots to be freed right near the end of the contract. The needed duration of the attack would depend on the proof frequency, however 30 days should be way more time than is needed.

I really like the idea of burning a percentage of the funds that would be returned to the client, however:
1. I see the cost of this attack as the mitigation needed, as there is no financial gain to be had from performing the attack. The only gain is disruption of the network. 
2. From an honest user's perspective, getting all funds returned in case the network fails to live up to its promises is a minimal level of consolation needed. If the user were to lose funds when the network fails (burned), I'm not sure how eager they'd be to come back and use Codex again.

Reply:
> 30 days is long enough for this attack. A DDOS attack would only need to be long enough that it can force slots to be freed right near the end of the contract.

I am talking here about the client's attack, which could "wiggle out" of the payment for the stored data if he DDOSed the providers at the end of the storage requests and hence got returned the full reward amount.

What I am mentioning here is that if the limit of storage request duration is low, then the price for the request is low, which I would argue that the gain from such an attack should be small enough to justify a DDOS attack on several node providers. 

ChatGPT (hence take it with a grain of salt) says the cheapest DDOS starts at 10$ per hour. Hence, you would need at least 24 hours for several days in order for the providers to get kicked out of the request and make it fail. So the cost to kick out a single provider starts at least on: `$10 * 24 * 4 = $960` (the 4 is the arbitrary number for how many slashes it takes before the node is kicked out). Based on the tolerance setting, there might need to be more providers kicked out as well.

Reply:
I'm pretty sure we're saying the same thing 😅  I was only trying to point out that our "low duration" of 30 days is already long enough to allow for DDOS attacks and so this is somewhat of a moot point in arguing for keeping contract duration low.

> the gain from such an attack
Technically, there is no gain except "free" storage and network disruption, as there is no flow of *additional* funds to the client. They would simply get their own funds returned to them. Well they might get reimbursed for the gas cost associated with creating the request, but there's no way that would exceed the cost of a DDOS attack. It'll almost definitely be cheaper to actually pay for the storage 😄